bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+        balance. Thus the probability of accepting a proposal is multiplied by the
+        hasting term of the propopsed sample divided by the hastings term of the
+        current sample.
+
+        Args:
+            hastings: log probability of rejecting all previous proposals
+
+        Returns:
+            float: log probability of making another proposal
+        """
+        return hastings if self._probabilistic else 0.0
+
+    def sample(self) -> DrawAndLogP:
+        """Sample from target distribution with DrGhmcDiag sampler.
+
+        From current sample (theta, rho), propose new sample (theta_prop, rho_prop)
+        and compute its acceptance probability. If accepted, return the new sample and
+        its log density; if rejected, propose a new sample and repeat. Stop when the
+        maximum number of proposals is reached or when probabilistic delayed rejection
+        terminates early.
+
+        Flip momemntum after each proposal to ensure detailed balance. Flip momentum
+        before return statement for momentum persistence in Generalized HMC.
+
+        Returns:
+            DrawAndLogP: Tuple of (sample, log density)
+        """
+        self._rho = self._rng.normal(
+            loc=self._rho * np.sqrt(1 - self._dampening),
+            scale=self._dampening,
+            size=self._dim,
+        )
+        logp_cur = self.joint_logp(self._theta, self._rho)
+        hastings_cur, reject_logp = 0.0, 0.0
+
+        for k in range(self._num_proposals):
+            pdr = self._prob_of_delayed_rejection(reject_logp)
+            if not np.log(self._rng.uniform()) < pdr:
+                break
+
+            stepsize, steps = self._stepsize_list[k], self._steps_list[k]
+            theta_prop, rho_prop = self.leapfrog(
+                self._theta, self._rho, stepsize, steps
+            )
+            rho_prop = -rho_prop
+
+            accept_logp, logp_prop = self.accept(
+                theta_prop, rho_prop, k, hastings_cur, logp_cur
+            )
+            if np.log(self._rng.uniform()) < accept_logp:
+                self._theta, self._rho = theta_prop, rho_prop
+                logp_cur = logp_prop
+                break
+
+            reject_logp = np.log1p(-np.exp(accept_logp))
+            hastings_cur += reject_logp
+        self._rho = -self._rho  # always flip momentum in GHMC
+        return self._theta, logp_cur
+
+    def accept(
+        self,
+        theta_prop: VectorType,
+        rho_prop: VectorType,
+        k: int,
+        hastings_cur: float,
+        logp_cur: float,
+    ) -> tuple[float, float]:
+        """Log acceptance probability of transitioning from current to proposed sample.
+
+        Recursively compute log probability density of accepting transition from
+        current sample (theta, rho) to proposed sample (theta_prop, rho_prop).
+        Information about the current sample is contained in logp_cur and is *not*
```

Comment:
You can just say it's included---you don't need to explain why in doc like this, especially when it's obvious.